core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -0,0 +1,239 @@
+    mIterator = mCache.entrySet().iterator();
+    mEvictLock = new ReentrantLock();
+    mLastSizeWarningTime = 0;
+  }
+
+  private void evictIfOverLimit() {
+    long numToEvict = mCache.size() - mSoftLimit;
+
+    if (numToEvict <= 0) {
+      return;
+    }
+
+    if (mEvictLock.tryLock()) {
+      try {
+        // This thread won the race as the evictor
+        while (numToEvict > 0) {
+
+          if (!mIterator.hasNext()) {
+            mIterator = mCache.entrySet().iterator();
+          }
+          Map.Entry<K, ValNode> candidateMapEntry = mIterator.next();
+          ValNode candidate = candidateMapEntry.getValue();
+          if (candidate.mIsAccessed) {
+            candidate.mIsAccessed = false;
+          } else {
+            if (candidate.mRefCount.compareAndSet(0, Integer.MIN_VALUE)) {
+              // the value object can be evicted, at the same time we make refCount minValue
+              mIterator.remove();
+              numToEvict--;
+            }
+          }
+        }
+      } finally {
+        mEvictLock.unlock();
+      }
+    }
+  }
+
+  /**
+   * If the key has an associated lock in the cache, return a LockResource for that key and
+   * the associated lock using the specified lockmode.  If the key has no entry in the cache,
+   * create a lock and return a LockResource for that lock.
+   *
+   * @param key the key to look up the cache
+   * @param mode lockMode to acquire
+   * @return the value contained in the cache, if it is already in cache,
+   *         otherwise generate an entry based on the loader
+   */
+  public LockResource get(K key, LockMode mode) {
+    ValNode valNode = getValNode(key);
+    ReentrantReadWriteLock lock = valNode.mValue;
+    switch (mode) {
+      case READ:
+        return new RefCountLockResource(lock.readLock(), valNode.mRefCount);
+      case WRITE:
+        return new RefCountLockResource(lock.writeLock(), valNode.mRefCount);
+      default:
+        throw new IllegalStateException("Unknown lock mode: " + mode);
+    }
+  }
+
+  /**
+   * Get the raw readwrite lock from the cache.
+   *
+   * @param key key to look up the value
+   * @return the lock associated with the key
+   */
+  @VisibleForTesting
+  public ReentrantReadWriteLock getRawReadWriteLock(K key) {
+    return mCache.getOrDefault(key, new ValNode(new ReentrantReadWriteLock())).mValue;
+  }
+
+  private ValNode getValNode(K key) {
+    Preconditions.checkNotNull(key, "key can not be null");
+    // oldCacheEntry keeps track of the last cache entry that was in the process of being removed.
+    // we do not need to do anything if we did not get a new CacheEntry
+    ValNode oldCacheEntry = null;
+    ValNode cacheEntry;
+    while (true) {
+      // repeat until we get a cacheEntry that is not in the process of being removed.
+      cacheEntry = mCache.compute(key, (k, v) -> {
+        if (v != null) {
+          v.mRefCount.incrementAndGet();
```

Comment:
here, can we check the count, and if it is `< 0`, return `null` to repeat?

Then, I think we can remove the second loop and the old entry variable, right?

Reply:
sounds good, i was hesitant to do that because returning `null` means to remove it from the map as well. I checked `ConcurrentHashMap` iterator implementation, it just calls `remove` which is thread safe.

Reply:
on a second thought, having multiple threads removing values is still dangerous, because we could do `remove`, then `add` then `remove` again, and return a value that is not stored in the map.